packages/oauth/src/index.ts

```diff
@@ -314,7 +327,7 @@ export class InstallProvider {
           code,
           client_id: this.clientId,
           client_secret: this.clientSecret,
-          redirect_uri: installOptions.redirectUri,
+          redirect_uri: installOptions ? installOptions.redirectUri : undefined,
```

Comment:
If an app disables state validation while using a redirect_uri, this API call may fail. right? We may want to add `installOptions` as an optional argument of `handleCallback` method for it (I haven't verified how it works well yet, though).

Reply:
@seratch From my read of things, it appears that installOptions is an object encoded in the `state` param. So I'm not sure how it could be passed to `handleCallback` on the HTTPReceiver side. Is that your understanding as well?

Reply:
@srajiang I haven't verified if this works yet but I was thinking that we can have the following change on the bolt-js side.

```diff
diff --git a/src/receivers/HTTPReceiver.ts b/src/receivers/HTTPReceiver.ts
index 71188f69..9114ec3c 100644
--- a/src/receivers/HTTPReceiver.ts
+++ b/src/receivers/HTTPReceiver.ts
@@ -435,12 +435,17 @@ export default class HTTPReceiver implements Receiver {
     // when installer is defined then installCallbackOptions is always defined too.
     const [installer, installCallbackOptions] = [this.installer!, this.installCallbackOptions!];
 
-    installer.handleCallback(req, res, installCallbackOptions).catch((err) => {
+    const errorHandler = (err: Error) => {
       this.logger.error(
         'HTTPReceiver encountered an unexpected error while handling the OAuth install redirect. Please report to the maintainers.',
       );
       this.logger.debug(`Error details: ${err}`);
-    });
+    };
+    if (installer.stateValidation) {
+      installer.handleCallback(req, res, installCallbackOptions).catch(errorHandler);
+    } else {
+      installer.handleCallback(req, res, installCallbackOptions, this.installUrlOptions).catch(errorHandler);
+    }
   }
 }
 ```

Reply:
I noticed that
1) `installUrlOptions` [doesn't contain ](https://github.com/slackapi/bolt-js/blob/ee18b1c7d7b5c80a0fcce66b5c48a8235f2417ec/src/receivers/HTTPReceiver.ts#L170-L174)`redirectUrl` field. This is the same object being [passed to](https://github.com/slackapi/bolt-js/blob/ee18b1c7d7b5c80a0fcce66b5c48a8235f2417ec/src/receivers/HTTPReceiver.ts#L464) `generateInstallUrl` and encoded as `state`. Then later on in handleCallback, we are trying to access redirectUri (for example [here](https://github.com/slackapi/node-slack-sdk/blob/66066ee4dfa069c4829bc239b33a98a87cb1c009/packages/oauth/src/index.ts#L351)) from the installOptions, but actually it's always undefined. This is a bug I can fix on the HTTPReceiver side.

2) `installUrlOptions` options object contains fields set during App initialization which are static (including even `metadata`) so they don't appear to change between installations. `installURLOptions` can always be available as a parameter to `handleCallback()`, and we could still verify `state` as planned (it throws an error if state is invalid). It's a different design than the original, but having options passed around as state seems to only make sense if options are dynamic. @stevengill re: our chat

Reply:
Alrighty, maybe 2. is too drastic of a change - I've made changes to support passing in installOptions to `handleCallback()`

Reply:
@srajiang Thanks for checking these! 

Regarding the point 1., we should fix the bug for all the built-in receivers.

As for 2., perhaps, we can revisit the issues around the metadata. A static value for metadata is not useful at all. We should consider adding a way to set metadata per request. Also, some developers may want to switch redirect_uri per requests in some cases. 

For your information, in Java / Python SDKs, I have been encouraging to use browser cookies to tell the OAuth callback (=redirect_uri) the state/user tracking data. Perhaps, something similar rather than embedding value in the state parameter would be more robust for any use cases.